bwscanner/attacher.py

```diff
@@ -156,30 +156,33 @@ def connect_to_tor(launch_tor, circuit_build_timeout, control_port=None,
     if tor_overrides:
         tor_options.update(tor_overrides)
 
+    tor_config = txtorcon.TorConfig()
+
+    # Update Tor config options from dictionary
+    for key, value in tor_options.items():
+        setattr(tor_config, key, value)
+
     if launch_tor:
         log.info("Spawning a new Tor instance.")
         # TODO: Pass in data_dir directory so consensus can be cached
-        tor = yield txtorcon.launch(reactor)
+        # Launch tor with config, in order to don't get CONF_CHANGED when
+        # updating options that can't be changed while tor is running.
+        tor = yield txtorcon.launch(reactor, _tor_config=tor_config)
     else:
         log.info("Trying to connect to a running Tor instance.")
         if control_port:
             endpoint = endpoints.TCP4ClientEndpoint(reactor, "localhost", control_port)
         else:
             endpoint = None
         tor = yield txtorcon.connect(reactor, endpoint)
+        # TODO: check whether CONF_CHANGED will happen here or not because
+        # we get the state later
+        tor.config = tor_config
+        tor.config.save()
```

Comment:
You shouldn't really assign to this in the `Tor` object. Nor should you have to `save()` (all config will have been passed in to tor on the command-line). If you do call `.get_config()` anywhere else, it will return the same object (but, you should use the async method to retrieve the config in case it's not available yet).

BTW, do you know which config-options you're passing in that aren't provided by `launch()`, or is it "maybe anything at all that Tor supports"..?

Reply:
The options we're passing are (https://github.com/juga0/bwscanner/blob/ccddac6f20ce7963e8fc2e8d9921f672da6be390/bwscanner/attacher.py#L146):
    # Options for spawned or running Tor to load the correct descriptors.
    tor_options = {
        'LearnCircuitBuildTimeout': 0,  # Disable adaptive circuit timeouts.
        'CircuitBuildTimeout': circuit_build_timeout,
        'UseEntryGuards': 0,  # Disable UseEntryGuards to avoid PathBias warnings.
        'UseMicroDescriptors': 0,
        'FetchUselessDescriptors': 1,
        'FetchDirInfoEarly': 1,
        'FetchDirInfoExtraEarly': 1,
    }

So what would be the correct way to force change the configuration (if any) when ``tor`` is already running (by command line) and we use ``txtorcon.connect``?

Reply:
From http://timaq4ygg2iegci7.onion/guide.html#connecting-to-a-running-tor:

    Similarly, the current configuration is available via .config. You can change the configuration
    by updating attributes on this class but it won’t take effect until you call txtorcon.TorConfig.save().

Reply:
Hmm, that does read confusingly now that I re-read it. What I *mean* there is that you can "update things" by e.g. `tor.config.UseEntryGuards = 1` and `tor.config.save()` not by setting `tor.config` directly. I'll try to make that docstring more-clear... :/

Reply:
This commit https://github.com/TheTorProject/bwscanner/pull/85/commits/353ebac4301c8e5a3038341afad9b626f6bdbe2b#diff-5d97531ed54d2f8c4d91ebce0fb91774R180 should fix it.

I also moved ``TorConfig`` setup inside of ``if launch_tor`` https://github.com/TheTorProject/bwscanner/commit/353ebac4301c8e5a3038341afad9b626f6bdbe2b#diff-5d97531ed54d2f8c4d91ebce0fb91774R163